### PR TITLE
exp: Unify dvc exp list and show output

### PR DESCRIPTION
**Pull request by tibor-mach**
Solves #9794

### CODE REVIEW

File: dvc/repo/experiments/ls.py [hunk 1]
```diff
@@ -31,18 +31,19 @@ def ls(
...
 
     results = defaultdict(list)
     for baseline in ref_info_dict:
         name = baseline
-        if tags[baseline] or ref_heads[baseline]:
-            name = tags[baseline] or ref_heads[baseline]
+        if baseline_names[baseline]:
+            # the str() retyping is a hack to deal with the typed dictionary
+            # from _describe...probably there is a better way
+            name = str(baseline_names.get(baseline))
```
Comment (tibor-mach) [inline dvc/repo/experiments/ls.py:46]: This can probably be done in a nicer way...I haven't worked with typing much, looking forward to any suggestions.

File: dvc/repo/experiments/utils.py [hunk 2]
```diff
@@ -342,3 +343,60 @@ def to_studio_params(dvc_params):
...
 
     return result
+
+
+def _describe(
+    scm: "Git",
+    revs: Iterable[str],
+    logger,
+    refs: Optional[Iterable[str]] = None,
+    remove_ref_string=True,
```
Comment (tibor-mach) [inline dvc/repo/experiments/utils.py:353]: This function is used for `dvc exp list` and `dvc exp show`. For some reason `dvc exp show` uses the shortened path without the `ref/...` and `dvc exp list` uses the full path. A couple of tests break unless the output is formatted differently for each of them.

I haven't looked too much into it but it seems that simply rewriting the simple test for `dvc exp list` output is not enough, there are other things affected if we just keep the original `_describe`. Might be worth refactoring as the regex probably makes things slower (judging by the GH tests).

File: dvc/repo/experiments/utils.py [hunk 3]
```diff
@@ -342,3 +343,63 @@ def to_studio_params(dvc_params):
             result[file_name] = file_data.get("data", {})
 
     return result
+
+
+def _describe(
```
Comment (pmrowla) [inline dvc/repo/experiments/utils.py:348]:
```suggestion
def describe(
```

Since this is now a public `experiments.utils` method we should drop the `_` prefix

File: dvc/repo/experiments/utils.py (hunk 2 above)
Comment (pmrowla) [reply dvc/repo/experiments/utils.py:353]: I think what we really want is to make `experiments.ls()` only return git SHAs (so we would not call `describe()` at all in `dvc.repo.experiments.ls()`. At the UI/command level we can use `describe()` to get the names without the prefix when we actually want them.

So in `dvc.repo.experiments.ls` we can drop all of the git describe `name` handling entirely, so the returned result:
https://github.com/iterative/dvc/blob/43b371c43a090a5c05344f93538c7114a150319b/dvc/repo/experiments/ls.py#L51
now always only uses the baseline git SHA as the dict keys like:
```python
results[baseline].append(...)
```

The human-readable name handling would then be done strictly at the UI level, in `dvc.commands.experiments.ls`

Basically, we would replace this section:
https://github.com/iterative/dvc/blob/43b371c43a090a5c05344f93538c7114a150319b/dvc/commands/experiments/ls.py#L26-L35

with something like
```python
if name_only or sha_only:
    names = {}
else:
    names = describe(baseline for baseline in exps)
for baseline in exps:
    if not (name_only or sha_only):
        name = names.get(baseline) or baseline[:7]
        ui.write(f"{name}:")
    ...
```

File: dvc/repo/experiments/utils.py [hunk 4]
```diff
@@ -342,3 +343,63 @@ def to_studio_params(dvc_params):
...
+            names[rev] = head_branch
+        else:
+            names[rev] = tags.get(rev) or branches.get(rev)
+
+    if remove_ref_string:
+        replace_pat = r"^refs/(?:tags|heads)/"
+        names = {
+            rev: (re.sub(replace_pat, "", name) if name else name)
+            for rev, name in names.items()
+        }
```
Comment (pmrowla) [inline dvc/repo/experiments/utils.py:403]: This can be removed (see my other comments)

**Review by pmrowla — CHANGES_REQUESTED** [review]
Thanks for the PR! It's the right idea, I left some comments on how we could make the internal APIs a little bit cleaner.

File: dvc/repo/experiments/utils.py (hunk 3 above)
Comment (tibor-mach) [reply dvc/repo/experiments/utils.py]: Makes sense, done.

**Comment by tibor-mach** [comment]
@pmrowla I made the changes that you suggested and I think it all works well now and `describe` can be simpler, without the extra argument (basically the same as it was before).

However, some of the tests are written to check things in `dvc.experiments.ls`which are now broken (since they check that the references are already returned by `dvc.experiments.ls`. Should I rewrite those tests? I guess it actually makes more sense to test the `endpoints` as it were, so to test at the level of `dvc.commands.ls`. I would probably need a bit of help there to understand the way tests are set up in dvc etc.

The tests in question were all written by @dberenbaum , so maybe we can have a look at that together (seeing as we are closer timezone-wise)

**Comment by pmrowla** [comment]
@tibor-mach the tests for `dvc.experiments.ls` can be updated as needed

**Comment by tibor-mach** [comment]
Ok, I will have a look at how the tests are done and if I need any help, I will ask @dberenbaum

**Comment by tibor-mach** [comment]
@pmrowla the tests should all work as intended now. I am not sure why

Also, shouldn't we have a test that checks the desired order of head-tag-branch on the UI level for both exp show and exp list? Or maybe it would be enough to test the newly public `describe` utility?

Also please have a look at how I handled the typing in the UI. Perhaps that is what is making it fail benchmarks? Thanks!

File: dvc/commands/experiments/ls.py [hunk 5]
```diff
@@ -23,15 +27,20 @@ def run(self):
             git_remote=git_remote,
         )
 
+        if name_only or sha_only:
+            names = {}
+        elif isinstance(self.repo.scm, NoSCM):
+            raise SCMError("No SCM repository found")
```
Comment (pmrowla) [inline dvc/commands/experiments/ls.py:33]: You don't need to check this here, we raise an exception as soon as `repo.experiments` is accessed if the repo is no-scm (so when we call `experiments.ls()`)

File: dvc/commands/experiments/ls.py [hunk 6]
```diff
@@ -1,9 +1,13 @@
 import argparse
 import logging
 
+from scmrepo.exceptions import SCMError
```
Comment (pmrowla) [inline dvc/commands/experiments/ls.py:4]: This is unneeded

Also, for future reference in DVC we prefer using `dvc.scm.SCMError` and not `scmrepo.exceptions.SCMError` (so that we end up raising an exception that derives from `DVCException`)

File: dvc/commands/experiments/ls.py [hunk 7]
```diff
@@ -1,9 +1,13 @@
...
+
 from dvc.cli.command import CmdBase
 from dvc.cli.utils import append_doc_link
 from dvc.exceptions import InvalidArgumentError
+from dvc.repo.experiments.utils import describe
+from dvc.scm import NoSCM
```
Comment (pmrowla) [inline dvc/commands/experiments/ls.py:10]: This is unneeded

File: dvc/commands/experiments/ls.py [hunk 8]
```diff
@@ -1,9 +1,13 @@
 import argparse
 import logging
 
+from scmrepo.exceptions import SCMError
+
 from dvc.cli.command import CmdBase
 from dvc.cli.utils import append_doc_link
 from dvc.exceptions import InvalidArgumentError
+from dvc.repo.experiments.utils import describe
```
Comment (pmrowla) [inline dvc/commands/experiments/ls.py:9]: This import should be lazy (so the import statement needs to be moved inside `CmdExperimentsList.run()` where we actually use `describe()`).

The benchmark CI failure is because importing `dvc.repo.experiments` here adds a lot of overhead to the default DVC initialization (we only want to import it when we are actually running an experiments command, and not on any other DVC command)

**Review by pmrowla — CHANGES_REQUESTED** [review]

File: dvc/commands/experiments/ls.py (hunk 8 above)
Comment (tibor-mach) [reply dvc/commands/experiments/ls.py:9]: @pmrowla Makes sense...in general what are the best practices for lazy vs eager imports? Why don't we actually use lazy imports everywhere (perhaps except for stuff that needs to be imported for every function/method/class in a module)? This is actually something I wondered about before with other projects.

Comment (pmrowla) [reply dvc/commands/experiments/ls.py]: It varies from project to project, but I think the general consensus for python is usually that you should not use lazy imports until you have a specific requirement to do so (like when you need to avoid having a circular import or for specific performance reasons)

**Comment by tibor-mach** [comment]
```
FAILED tests/func/test_update.py::test_update_import_url_to_remote_directory - PermissionError: [WinError 5] Access is denied: 'C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pytest-of-runneradmin\\pytest-0\\popen-gw0\\local-cloud96\\cache\\files\\md5\\ac\\.DGsve8RXEpR2pWbAyys8Qp.tmp' -> 'C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pytest-of-runneradmin\\pytest-0\\popen-gw0\\local-cloud96\\cache\\files\\md5\\ac\\bd18db4cc2f85cedef654fccc4a4d8'
```

I'm a bit confused about this error on Windows. It only seems to happen with Python 3.9 and it doesn't really seem to be related to anything I changed...

@pmrowla any ideas?

**Comment by pmrowla** [comment]
It's a known flaky test, you can disregard that failure since it's not related to your changes

**Review by pmrowla — APPROVED** [review]
